### PR TITLE
cli: Cancelable queries and graceful shutdown

### PR DESCRIPTION
**Pull request by asdine**
This PR adds support for cancelable commands and queries, and graceful shutdown by changing the way the command line tool handles user input and execution.

Major changes:
- Graceful shutdown: Errors and termination bubble up to the main, no `os.Exit` calls from commands.
- Ctrl D on an empty line triggers graceful shutdown
- Ctrl C / SIGINT cancels running commands
- SIGTERM (i.e. `pkill genji`) triggers graceful shutdown

Subsequent PRs will simplify the code further so that it will be easier to add tests.

### CODE REVIEW

File: cmd/genji/shell/shell.go [hunk 1]
```diff
@@ -19,12 +23,23 @@ import (
 	"github.com/genjidb/genji/engine/boltengine"
 	"github.com/genjidb/genji/engine/memoryengine"
 	"github.com/genjidb/genji/sql/parser"
+	"github.com/hashicorp/go-multierror"
```
Comment (tie) [inline cmd/genji/shell/shell.go:26]: YMMV but I think go.uber.org/multierr has a better API design.

**Review by tie — COMMENTED** [review]
Neat!

File: cmd/genji/shell/shell.go [hunk 2]
```diff
@@ -103,25 +125,144 @@ func Run(opts *Options) error {
...
+}
+
+// runSignalHandlers handles two different signals.
+// On SIGINT, it cancels any query execution using sh.cancelExecution.
+// On SIGTERM, it triggers graceful shutdown by returning errExitSignal.
+func (sh *Shell) runSignalHandlers(ctx context.Context) error {
+	interruptC := make(chan os.Signal, 1)
+	termC := make(chan os.Signal, 1)
+	signal.Notify(interruptC, os.Interrupt)
+	signal.Notify(termC, syscall.SIGTERM)
```
[the remark this answers is not in the document]
Comment (asdine) [reply cmd/genji/shell/shell.go:198]: Indeed, how would that work for other systems?

File: cmd/genji/shell/shell.go (hunk 1 above)
Comment (asdine) [reply cmd/genji/shell/shell.go:26]: Sure, I'll take a look

File: cmd/genji/shell/shell.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tie) [reply cmd/genji/shell/shell.go:198]: Well, it used to be undefined on some systems (e.g. js/wasm, https://github.com/golang/go/issues/28719) or be a no-op, e.g. on Windows we had to handle different signal before go1.14, see https://github.com/golang/go/issues/7479 and https://stackoverflow.com/a/47314406.

Looking at `grep SIGTERM` output in `syscall` package, it seems to be defined for all major all targets now. So perhaps it should be safe to use `SIGTERM` despite the [Go Proverbs](https://go-proverbs.github.io): [Syscall must always be guarded with build tags](https://youtu.be/PAAkCSZUG1c?t=11m10s).

**Pull request merged**